pkg/sources/forward/data_forward.go

```diff
@@ -204,6 +211,31 @@ func (isdf *DataForward) forwardAChunk(ctx context.Context) {
 		metrics.ReadMessagesError.With(map[string]string{metrics.LabelVertex: isdf.vertexName, metrics.LabelPipeline: isdf.pipelineName, metrics.LabelVertexType: string(dfv1.VertexTypeSource), metrics.LabelVertexReplicaIndex: strconv.Itoa(int(isdf.vertexReplica)), metrics.LabelPartitionName: isdf.reader.GetName()}).Inc()
 	}
 
+	// execute only if idle source configuration is defined, i.e.
+	// maxWait: How much longer it will wait before increasing the idle watermark
+	// minIncrement: Duration to increase in case of watermark idling
+	if isdf.Watermark.IdleSource != nil {
+		isWMIdle := isdf.isWatermarkIdle()
+
+		// If no any message is coming and watermark is idling then start increasing the watermark with "minIncrement"
+		// value after each "maxWait" duration
+		if len(readMessages) == 0 && isWMIdle {
+			minIncrement := isdf.Watermark.IdleSource.GetMinIncrement()
+			// Get the head watermark value then add the minIncrement to it and publish the idle watermark with updated value.
+			headWatermark := isdf.wmFetcher.ComputeHeadWatermark(0)
+			updatedWM := headWatermark.Add(minIncrement)
+
+			isdf.srcWMPublisher.PublishIdleWatermarks(updatedWM)
+			// set the time for last updated watermark which will be used for calculating maximum wait time is passed
+			// or not in case of idle watermark.
+			isdf.lastUpdateWM = time.Now()
+			// currentWatermark is used to calculate when watermark is idling
+			isdf.currentWatermark = isdf.wmFetcher.ComputeWatermark(nil, 0)
```

Comment:
add a comment about `isWatermarkIdle`, also move this inside the if check, in line 241